File: src/undef/telemetry/pii.py
# SPDX-FileCopyrightText: Copyright (C) 2026 MindTenet LLC
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of Undef Telemetry.
#

"""PII policy engine with nested traversal support."""

from __future__ import annotations

import copy
import hashlib
import threading
from dataclasses import dataclass
from typing import Any, Literal

MaskMode = Literal["drop", "redact", "hash", "truncate"]


@dataclass(frozen=True, slots=True)
class PIIRule:
    path: tuple[str, ...]
    mode: MaskMode = "redact"
    truncate_to: int = 8


_DEFAULT_SENSITIVE_KEYS = {"password", "token", "authorization", "api_key", "secret"}
_lock = threading.Lock()
_rules: list[PIIRule] = []


def replace_pii_rules(rules: list[PIIRule]) -> None:
    with _lock:
        _rules.clear()
        _rules.extend(rules)


def register_pii_rule(rule: PIIRule) -> None:
    with _lock:
        _rules.append(rule)


def get_pii_rules() -> tuple[PIIRule, ...]:
    with _lock:
        return tuple(_rules)


_REDACTED = "***"
_TRUNCATION_SUFFIX = "..."


def _mask(value: Any, mode: MaskMode, truncate_to: int) -> Any:
    if mode == "drop":
        return None
    if mode == "redact":
        return _REDACTED
    if mode == "hash":
        return hashlib.sha256(str(value).encode("utf-8")).hexdigest()[:12]
    text = str(value)
    limit = max(0, truncate_to)
    if len(text) <= limit:
        return text
    return text[:limit] + _TRUNCATION_SUFFIX


def _match(path: tuple[str, ...], target: tuple[str, ...]) -> bool:
    if len(path) != len(target):
        return False
    return all(part == "*" or part == elem for part, elem in zip(path, target, strict=True))


def _apply_rule(node: Any, rule: PIIRule, current_path: tuple[str, ...] = ()) -> Any:
    if isinstance(node, dict):
        output: dict[str, Any] = {}
        for key, value in node.items():
            child_path = (*current_path, key)
            if _match(rule.path, child_path):
                masked = _mask(value, rule.mode, rule.truncate_to)
                if masked is not None:
                    output[key] = masked
            else:
                output[key] = _apply_rule(value, rule, child_path)
        return output
    if isinstance(node, list):
        return [_apply_rule(item, rule, (*current_path, "*")) for item in node]
    return node


def _apply_default_sensitive_key_redaction(node: Any) -> Any:
    if isinstance(node, dict):
        output: dict[str, Any] = {}
        for key, value in node.items():
            if key.lower() in _DEFAULT_SENSITIVE_KEYS:
                if key in rule_targeted_keys or value != orig_value:
                    output[key] = value
                else:
                    output[key] = _REDACTED
            else:
                output[key] = _apply_default_sensitive_key_redaction(value)
        return output
    if isinstance(node, list):
        return [_apply_default_sensitive_key_redaction(item) for item in node]
    return node


def _collect_rule_leaf_keys(rules: tuple[PIIRule, ...]) -> frozenset[str]:
    """Collect the leaf key names that custom rules target."""
    return frozenset(rule.path[-1] for rule in rules if rule.path)


def _needs_deep_copy(rules: tuple[PIIRule, ...]) -> bool:  # pragma: no mutate
    """Return True if any rule targets a nested path (depth > 1)."""
    return any(len(rule.path) > 1 for rule in rules)  # pragma: no mutate


def sanitize_payload(payload: dict[str, Any], enabled: bool) -> dict[str, Any]:
    if not enabled:
        return dict(payload)
    rules = get_pii_rules()
    cleaned: Any = copy.deepcopy(payload) if _needs_deep_copy(rules) else dict(payload)  # pragma: no mutate
    for rule in rules:
        cleaned = _apply_rule(cleaned, rule)
    if isinstance(cleaned, dict):
        return cleaned
    return {}


def reset_pii_rules_for_tests() -> None:
    replace_pii_rules([])

File: go/logger/pii.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package logger

import (
	"crypto/sha256"
	"fmt"
	"regexp"
	"strings"
	"sync"
)

// PIIRule defines a rule for sanitizing a specific field path.
type PIIRule struct {
	Path       []string
	Mode       string
	TruncateTo int
}

// PII mode constants.
const (
	PIIModeRedact   = "redact"
	PIIModeDrop     = "drop"
	PIIModeHash     = "hash"
	PIIModeTruncate = "truncate"
)

const (
	_piiRedacted         = "***"
	_piiTruncationSuffix = "..."
	_piiDefaultMax       = 8
)

// _defaultSensitiveKeys lists case-insensitive exact-match key names.
var _defaultSensitiveKeys = map[string]struct{}{
	"password":       {},
	"passwd":         {},
	"secret":         {},
	"token":          {},
	"api_key":        {},
	"apikey":         {},
	"auth":           {},
	"authorization":  {},
	"credential":     {},
	"private_key":    {},
	"ssn":            {},
	"credit_card":    {},
	"creditcard":     {},
	"cvv":            {},
	"pin":            {},
	"account_number": {},
	"cookie":         {},
}

// SecretPattern pairs a diagnostic name with a compiled regexp.
type SecretPattern struct {
	Name    string
	Pattern *regexp.Regexp
}

var (
	_piiMu              sync.RWMutex
	_piiRules           []PIIRule
	_classificationHook func(key string, value any) string
	_receiptHook        func(fieldPath string, action string, originalValue any)
	_customSecretPats   map[string]*regexp.Regexp
	_sanitizeDelegate   func(map[string]any, bool, int) map[string]any
)

// RegisterSecretPattern registers a custom secret detection pattern.
func RegisterSecretPattern(name string, pattern *regexp.Regexp) {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	if _customSecretPats == nil { // pragma: allowlist secret
		_customSecretPats = make(map[string]*regexp.Regexp) // pragma: allowlist secret
	}
	_customSecretPats[name] = pattern
}

// GetSecretPatterns returns all secret patterns (built-in + custom).
func GetSecretPatterns() []SecretPattern {
	_piiMu.RLock()
	defer _piiMu.RUnlock()
	out := make([]SecretPattern, 0, len(_secretPatterns)+len(_customSecretPats))
	for i, re := range _secretPatterns {
		out = append(out, SecretPattern{Name: fmt.Sprintf("builtin-%d", i), Pattern: re})
	}
	for name, re := range _customSecretPats {
		out = append(out, SecretPattern{Name: name, Pattern: re})
	}
	return out
}

// ResetSecretPatterns clears all custom secret patterns (for test cleanup).
func ResetSecretPatterns() {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	_customSecretPats = nil // pragma: allowlist secret
}

// SetClassificationHook registers a classification callback on the PII engine.
func SetClassificationHook(fn func(string, any) string) {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	_classificationHook = fn
}

// SetReceiptHook registers a redaction receipt callback on the PII engine.
func SetReceiptHook(fn func(string, string, any)) {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	_receiptHook = fn
}

// SetPIIRules replaces the global PII rule list.
func SetPIIRules(rules []PIIRule) {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	cp := make([]PIIRule, len(rules))
	copy(cp, rules)
	_piiRules = cp
}

// GetPIIRules returns a copy of the current global PII rules.
func GetPIIRules() []PIIRule {
	_piiMu.RLock()
	defer _piiMu.RUnlock()
	cp := make([]PIIRule, len(_piiRules))
	copy(cp, _piiRules)
	return cp
}

// ResetPIIRules clears all custom PII rules, hooks, and the sanitize delegate (for test cleanup).
func ResetPIIRules() {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	_piiRules = nil
	_classificationHook = nil
	_receiptHook = nil
	_sanitizeDelegate = nil
}

// SetSanitizePayloadFunc registers an external sanitize function that SanitizePayload
// will delegate to instead of using its own rule set. Pass nil to deregister and
// fall back to the local rule set. This allows the top-level telemetry package to
// wire its own PII engine into the logger sub-package without creating an import cycle.
func SetSanitizePayloadFunc(fn func(map[string]any, bool, int) map[string]any) {
	_piiMu.Lock()
	defer _piiMu.Unlock()
	_sanitizeDelegate = fn
}

// SanitizePayload applies PII sanitization to the given payload map.
// If a delegate function has been registered via SetSanitizePayloadFunc, it is
// called instead of the local rule set, allowing the top-level telemetry engine
// to serve as the single source of truth for PII rules.
// If enabled is false, a shallow copy is returned unchanged.
func SanitizePayload(payload map[string]any, enabled bool, maxDepth int) map[string]any {
	if !enabled {
		return _shallowCopy(payload)
	}
	_piiMu.RLock()
	delegate := _sanitizeDelegate
	_piiMu.RUnlock()
	if delegate != nil {
		return delegate(payload, enabled, maxDepth)
	}
	if maxDepth <= 0 {
		maxDepth = _piiDefaultMax
	}
	rules := GetPIIRules()
	result := _sanitizeMap(payload, []string{}, rules, maxDepth)
	_piiMu.RLock()
	classHook := _classificationHook
	_piiMu.RUnlock()
	if classHook != nil {
		for k, v := range result {
			if label := classHook(k, v); label != "" {
				result["__"+k+"__class"] = label
			}
		}
	}
	return result
}

func _shallowCopy(m map[string]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func _sanitizeMap(m map[string]any, path []string, rules []PIIRule, depth int) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		childPath := append(path, k) //nolint:gocritic
		sanitized, drop := _sanitizeValue(k, v, childPath, rules, depth)
		if !drop {
			out[k] = sanitized
		}
	}
	return out
}

func _sanitizeSlice(s []any, path []string, rules []PIIRule, depth int) []any {
	out := make([]any, 0, len(s))
	for _, item := range s {
		if inner, ok := item.(map[string]any); ok {
			out = append(out, _sanitizeMap(inner, path, rules, depth))
		} else {
			sanitized, drop := _sanitizeValue("", item, path, rules, depth)
			if !drop {
				out = append(out, sanitized)
			}
		}
	}
	return out
}

func _fireReceiptHook(hook func(string, string, any), fieldPath, action string, original any) {
	if hook != nil {
		hook(fieldPath, action, original)
	}
}

func _sanitizeValue(key string, value any, path []string, rules []PIIRule, depth int) (any, bool) {
	_piiMu.RLock()
	receiptHook := _receiptHook
	_piiMu.RUnlock()
	for _, rule := range rules {
		if _applyRule(rule, path) {
			_fireReceiptHook(receiptHook, strings.Join(path, "."), rule.Mode, value)
			return _applyMode(value, rule.Mode, rule.TruncateTo)
		}
	}
	if _isDefaultSensitiveKey(key) {
		_fireReceiptHook(receiptHook, key, PIIModeRedact, value)
		return _piiRedacted, false
	}
	if str, ok := value.(string); ok && _detectSecretInValue(str) {
		_fireReceiptHook(receiptHook, key, PIIModeRedact, value)
		return _piiRedacted, false
	}
	if depth <= 1 {
		return value, false
	}
	switch typed := value.(type) {
	case map[string]any:
		return _sanitizeMap(typed, path, rules, depth-1), false
	case []any:
		return _sanitizeSlice(typed, path, rules, depth-1), false
	}
	return value, false
}

func _applyRule(rule PIIRule, path []string) bool {
	if len(rule.Path) != len(path) {
		return false
	}
	for i, seg := range rule.Path {
		if seg != "*" && seg != path[i] {
			return false
		}
	}
	return true
}

func _applyMode(value any, mode string, truncateTo int) (any, bool) {
	switch mode {
	case PIIModeDrop:
		return nil, true
	case PIIModeHash:
		sum := sha256.Sum256([]byte(fmt.Sprintf("%v", value)))
		return fmt.Sprintf("%x", sum)[:12], false
	case PIIModeTruncate:
		s := fmt.Sprintf("%v", value)
		runes := []rune(s)
		if len(runes) >= truncateTo+1 {
			return string(runes[:truncateTo]) + _piiTruncationSuffix, false
		}
		return s, false
	default:
		return _piiRedacted, false
	}
}

func _isDefaultSensitiveKey(key string) bool {
	_, ok := _defaultSensitiveKeys[strings.ToLower(key)]
	return ok
}

const _minSecretLength = 20

var _secretPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?:AKIA|ASIA)[A-Z0-9]{16}`),
	regexp.MustCompile(`eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}`),
	regexp.MustCompile(`gh[pos]_[A-Za-z0-9_]{36,}`),
	regexp.MustCompile(`[0-9a-fA-F]{40,}`),
	regexp.MustCompile(`[A-Za-z0-9+/]{40,}={0,2}`),
}

func _detectSecretInValue(s string) bool {
	if len(s) < _minSecretLength {
		return false
	}
	for _, re := range _secretPatterns {
		if re.MatchString(s) {
			return true
		}
	}
	_piiMu.RLock()
	customs := _customSecretPats
	for _, re := range customs {
		if re.MatchString(s) {
			_piiMu.RUnlock()
			return true
		}
	}
	_piiMu.RUnlock()
	return false
}
